### PR TITLE
Rolled back, improved and cleaned-up the ZIP Stream

### PR DESCRIPTION
**Pull request by esiragusa**

### CODE REVIEW

File: include/seqan/stream/iostream_bgzf.h [hunk 1]
```diff
@@ -924,27 +871,26 @@ class basic_bgzf_istream :
...
         m_is_gzip(false),
         m_gbgzf_data_size(0)
     {};
 
-    /// returns true if it is a gzip file
-    bool is_gzip() const                {    return m_is_gzip;};
-    /// return data size check
-    bool check_data_size() const        {    return this->get_out_size() == m_gbgzf_data_size;};
+    // returns true if it is a gzip file
+    bool is_gzip() const                { m_is_gzip; };
```
Comment (rrahn) [inline include/seqan/stream/iostream_bgzf.h]: Missing return?

File: include/seqan/stream/iostream_bgzf.h [hunk 2]
```diff
@@ -924,27 +871,26 @@ class basic_bgzf_istream :
...
+    // return data size check
+    bool check_data_size() const        { this->get_out_size() == m_gbgzf_data_size; };
```
Comment (rrahn) [inline include/seqan/stream/iostream_bgzf.h]: Missing return here as well?

File: include/seqan/stream/iostream_bgzf.h [hunk 3]
```diff
@@ -924,27 +871,26 @@ class basic_bgzf_istream :
...
+
+    // return the data size in the file
+    long get_gbgzf_data_size() const    { m_gbgzf_data_size; };
```
Comment (rrahn) [inline include/seqan/stream/iostream_bgzf.h]: And here.

File: include/seqan/stream/iostream_bgzf.h [hunk 4]
```diff
@@ -815,44 +795,26 @@ class basic_bgzf_istreambase : virtual public std::basic_ios<Elem,Tr>
...
-    /// returns the compressed data size
-    long get_in_size() const                {    return m_buf.get_in_size();};
+    // returns the bgzf error state
+    int get_zerr() const                    { m_buf.get_zerr(); };
+    // returns the uncompressed data crc
+    long get_crc() const                    { m_buf.get_crc(); };
+    // returns the uncompressed data size
+    long get_out_size() const               { m_buf.get_out_size(); };
+    // returns the compressed data size
+    long get_in_size() const                { m_buf.get_in_size(); };
```
Comment (rrahn) [inline include/seqan/stream/iostream_bgzf.h]: Missing return statement for all functions above.

File: include/seqan/stream/iostream_zip_impl.h [hunk 5]
```diff
@@ -29,645 +13,530 @@
...
-            *pptr() = c;
-            ++w;
-        }
-        if ( zip_to_stream( pbase(), w ) ) {
-            this->setp( pbase(), epptr() );
-            return traits_type::not_eof( c );
-        } else {
-            return traits_type::eof();
-        }
+        this->setp(this->pbase(), this->epptr() - 1);
```
Comment (rrahn) [inline include/seqan/stream/iostream_zip_impl.h]: Why is epptr of the output buffer shrinked  by one?

**Comment by rrahn** [comment]
Ok except of the missing return values it all seems to be quite fine.

File: include/seqan/stream/iostream_bgzf.h [hunk 6]
```diff
@@ -800,14 +786,8 @@ template<
...
-        Tr,
-        ElemA,
-        ByteT,
-        ByteAT
-        > unbgzf_streambuf_type;
+    typedef std::basic_istream<Elem, Tr>&                           istream_reference;
+    typedef basic_unbgzf_streambuf<Elem, Tr, ElemA, ByteT, ByteAT>  unbgzf_streambuf_type;
 
     basic_bgzf_istreambase(istream_reference ostream_)
         : m_buf(ostream_)
```
Comment (h-2) [inline include/seqan/stream/iostream_bgzf.h]: move colon into line above

File: include/seqan/stream/iostream_zip.h [hunk 7]
```diff
@@ -64,88 +38,80 @@ enum EStrategy
...
+    // returns the size (bytes) of the input data compressed so far.
+    uLong get_in_size() const               { return m_zip_stream.total_in; }
 
-    /// @brief returns the size (bytes) of the input data compressed so far
-    uLong get_in_size() const { return m_zip_stream.total_in; }
-
-    /// @brief returns the size (bytes) of the compressed data so far
-    uLong get_out_size() const { return m_zip_stream.total_out; }
+    // returns the size (bytes) of the compressed data so far.
+    uLong get_out_size() const              { return m_zip_stream.total_out; }
```
Comment (h-2) [inline include/seqan/stream/iostream_zip.h]: what are these uLong types? Use __uint32/__uint64 instead? Or use standardized abbreviation instead...

File: include/seqan/stream/iostream_zip.h [hunk 8]
```diff
@@ -232,408 +188,278 @@ class basic_unzip_streambuf :
...
-            m_zip_stream_finalized = false;
-            return true;
-        }
-        return false;
-    }
-
-    static void put_long_as_uint32( ostream_reference out_, unsigned long x_ );
-
+    void put_uint32(unsigned long x_);
     void add_header();
```
Comment (h-2) [inline include/seqan/stream/iostream_zip.h]: either use fixed size integer types or not, but don't mix. unsigned long is 64bit on 64bit linux and mac!

File: include/seqan/stream/iostream_zip_impl.h [hunk 9]
```diff
@@ -29,645 +13,530 @@
...
+    m_crc(0)
+{
+    m_zip_stream.zalloc = (alloc_func)0;
+    m_zip_stream.zfree = (free_func)0;
+
+    m_zip_stream.next_in = NULL;
+    m_zip_stream.avail_in = 0;
+    m_zip_stream.avail_out = 0;
+    m_zip_stream.next_out = NULL;
+
```
Comment (h-2) [inline include/seqan/stream/iostream_zip_impl.h]: [not important] Why mix 0 and NULL?

File: include/seqan/stream/iostream_zip_impl.h [hunk 10]
```diff
@@ -29,645 +13,530 @@
...
+          typename Tr,
+          typename ElemA,
+          typename ByteT,
+          typename ByteAT>
+basic_unzip_streambuf<Elem, Tr, ElemA, ByteT, ByteAT>::basic_unzip_streambuf(
+    istream_reference istream_,
+    size_t window_size_,
+    size_t read_buffer_size_,
+    size_t input_buffer_size_
+    ) :
```
Comment (h-2) [inline include/seqan/stream/iostream_zip_impl.h]: move `) :` to line above?

also, are the sizes const? If yes, can we mark them as such?

**Comment by esiragusa** [comment]
Re-squashing.

**Pull request merged**